ext/opentelemetry-exporter-cloud-trace/tests/test_cloud_trace_propagator.py

```diff
@@ -0,0 +1,123 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import typing
+import unittest
+
+import opentelemetry.trace as trace
+from opentelemetry.context import get_current
+from opentelemetry.exporter.cloud_trace.cloud_trace_propagator import (
+    _TRACE_CONTEXT_HEADER_NAME,
+    CloudTraceFormatPropagator,
+)
+from opentelemetry.trace.span import (
+    SpanContext,
+    TraceFlags,
+    get_hexadecimal_trace_id,
+)
+
+
+def get_dict_value(dict_object: typing.Dict[str, str], key: str) -> str:
+    return dict_object.get(key, "")
+
+
+class TestCloudTraceFormatPropagator(unittest.TestCase):
+    def setUp(self):
+        self.propagator = CloudTraceFormatPropagator()
+        self.valid_trace_id = 281017822499060589596062859815111849546
+        self.valid_span_id = 17725314949316355921
+
+    def _extract(self, header_value):
+        """Test helper"""
+        header = {_TRACE_CONTEXT_HEADER_NAME: header_value}
+        new_context = self.propagator.extract(get_dict_value, header)
+        return trace.get_current_span(new_context).get_context()
+
+    def _inject(self, span=None):
+        """Test helper"""
+        ctx = get_current()
+        if span:
+            ctx = trace.set_span_in_context(span, ctx)
+        output = {}
+        self.propagator.inject(dict.__setitem__, output, context=ctx)
+        return output.get(_TRACE_CONTEXT_HEADER_NAME)
+
+    def test_no_context_header(self):
+        header = {}
+        new_context = self.propagator.extract(get_dict_value, header)
+        self.assertEqual(
+            trace.get_current_span(new_context).get_context(),
+            trace.INVALID_SPAN.get_context(),
+        )
+
+    def test_empty_context_header(self):
+        header = ""
+        self.assertEqual(
+            self._extract(header), trace.INVALID_SPAN.get_context()
+        )
+
+    def test_valid_header(self):
+        header = "{}/{};o=".format(
+            get_hexadecimal_trace_id(self.valid_trace_id), self.valid_span_id
+        )
+        new_span_context = self._extract([header])
+        self.assertEqual(new_span_context.trace_id, self.valid_trace_id)
+        self.assertEqual(new_span_context.span_id, self.valid_span_id)
+        self.assertEqual(new_span_context.trace_flags, TraceFlags(1))
+        self.assertTrue(new_span_context.is_remote)
+
+    def test_invalid_header(self):
+        header = "invalid_header"
+        self.assertEqual(
+            self._extract(header), trace.INVALID_SPAN.get_context()
+        )
+
+    def test_invalid_span_id(self):
+        header = "{}/{};o={}".format(
+            get_hexadecimal_trace_id(self.valid_trace_id), "0" * 16, 1
```

Comment:
Since the span ID is encoded as a decimal number, any digit string parsing as zero would be invalid - we should probably test that as well (e.g. single zero).

Reply:
SpanID of 0 is valid though? https://cloud.google.com/trace/docs/setup#force-trace
> SPAN_ID is the decimal representation of the (unsigned) span ID. It should be 0 for the first span in your trace. For subsequent requests, set SPAN_ID to the span ID of the parent request.

Reply:
Doesn't sound right - I'll look into that.